File: app/schemas/history_asset.py
from pydantic import BaseModel, Field
from typing import Any, Literal
from datetime import datetime


class HistoryAssetUploadResponse(BaseModel):
    id: int
    project_id: int
    asset_type: str
    original_filename: str | None
    parse_status: str
    case_count: int
    created_at: datetime

    model_config = {"from_attributes": True}


class HistoryAssetDetailResponse(BaseModel):
    id: int
    project_id: int
    asset_type: str
    original_filename: str | None
    parse_status: str
    parse_error: str | None
    parsed_cases: list[dict[str, Any]]
    case_count: int
    batch_id: int | None
    created_at: datetime
    updated_at: datetime | None

    model_config = {"from_attributes": True}


class HistoryClassificationItem(BaseModel):
    client_id: str = Field(..., min_length=1, max_length=80)
    classification: Literal["REUSE_CASE", "UPDATE_CASE", "NEW_CASE", "DEPRECATED_CASE", "CONFIRM_REQUIRED"]
    confidence: float = Field(1.0, ge=0.0, le=1.0)
    history_case: dict[str, Any] | None = None
    suggested_case: dict[str, Any] | None = None
    diff_fields: dict[str, Any] | None = None
    reason: str = ""
    matched_system_case_id: int | None = None


class HistoryClassificationSummary(BaseModel):
    reuse_count: int = 0
    update_count: int = 0
    new_count: int = 0
    deprecated_count: int = 0
    confirm_required_count: int = 0
    total: int = 0


class HistoryClassificationResponse(BaseModel):
    summary: HistoryClassificationSummary
    items: list[HistoryClassificationItem]

File: src/api/historyAsset.ts
import request from '@/utils/request'

export interface HistoryAssetItem {
  id: number
  project_id: number
  asset_type: 'excel' | 'xmind' | 'system_cases'
  original_filename: string | null
  parse_status: 'pending' | 'parsing' | 'completed' | 'failed'
  parse_error: string | null
  case_count: number
  batch_id: number | null
  created_at: string
  updated_at: string | null
}

export interface HistoryParsedCase {
  title: string
  module: string
  precondition: string
  steps: { action: string; expected_result: string }[]
  expected_result: string
  priority: number
  case_type: string
}

export interface HistoryAssetDetail extends HistoryAssetItem {
  parsed_cases: HistoryParsedCase[]
}

export type HistoryClassification = 'REUSE_CASE' | 'UPDATE_CASE' | 'NEW_CASE' | 'DEPRECATED_CASE' | 'CONFIRM_REQUIRED'

export interface HistoryClassificationItem {
  client_id: string
  classification: HistoryClassification
  confidence: number
  history_case: HistoryParsedCase | null
  suggested_case: HistoryParsedCase | null
  diff_fields: Record<string, { old: unknown; new: unknown }> | null
  reason: string
  matched_system_case_id: number | null
}

export interface HistoryClassificationSummary {
  reuse_count: number
  update_count: number
  new_count: number
  deprecated_count: number
  confirm_required_count: number
  total: number
}

export interface HistoryClassificationResponse {
  summary: HistoryClassificationSummary
  items: HistoryClassificationItem[]
}

function extractData<T>(response: unknown): T {
  const resp = response as { code?: number; data?: T }
  if (resp && resp.data !== undefined) return resp.data
  return response as T
}

export const historyAssetApi = {
  upload: async (projectId: number, file: File, assetType: string): Promise<HistoryAssetItem> => {
    const formData = new FormData()
    formData.append('file', file)
    formData.append('project_id', String(projectId))
    formData.append('asset_type', assetType)
    const response = await request.post('/api/v1/history-assets/upload', formData, {
      headers: { 'Content-Type': 'multipart/form-data' },
    })
    return extractData<HistoryAssetItem>(response)
  },

  getList: async (projectId: number): Promise<HistoryAssetItem[]> => {
    const response = await request.get('/api/v1/history-assets', { params: { project_id: projectId } })
    return extractData<HistoryAssetItem[]>(response)
  },

  getDetail: async (assetId: number): Promise<HistoryAssetDetail> => {
    const response = await request.get(`/api/v1/history-assets/${assetId}`)
    return extractData<HistoryAssetDetail>(response)
  },

  align: async (params: {
    project_id: number
    history_asset_ids: number[]
    requirement_file_ids?: number[]
    ui_screen_ids?: number[]
  }): Promise<HistoryClassificationResponse> => {
    const formData = new FormData()
    formData.append('project_id', String(params.project_id))
    formData.append('history_asset_ids', JSON.stringify(params.history_asset_ids))
    if (params.requirement_file_ids) formData.append('requirement_file_ids', JSON.stringify(params.requirement_file_ids))
    if (params.ui_screen_ids) formData.append('ui_screen_ids', JSON.stringify(params.ui_screen_ids))
    const response = await request.post('/api/v1/history-assets/align', formData)
    return extractData<HistoryClassificationResponse>(response)
  },

  importSystemCases: async (projectId: number, caseIds: number[]): Promise<HistoryAssetItem> => {
    const formData = new FormData()
    formData.append('project_id', String(projectId))
    formData.append('case_ids', JSON.stringify(caseIds))
    const response = await request.post('/api/v1/history-assets/import-system-cases', formData)
    return extractData<HistoryAssetItem>(response)
  },
}
